nav2_safety_nodes/src/nav2_safety_node.cpp

```diff
@@ -0,0 +1,250 @@
+  scan_subscribers_.reserve(scan_topics_.size());
+  RCLCPP_INFO(logger_, "Subscribing to scan topics");
+  for (unsigned int i = 0; i < scan_topics_.size(); i++) {
+    scan_subscribers_.push_back(
+      create_subscription<sensor_msgs::msg::LaserScan>(
+        scan_topics_[i].c_str(), rclcpp::SystemDefaultsQoS(),
+        std::bind(&SafetyZone::laserCallback, this, std::placeholders::_1)));
+  }
+  // Timer -> 10hzs
+  timer_ = create_wall_timer(
+    100ms, std::bind(&SafetyZone::timerCallback, this));
+  RCLCPP_INFO(logger_, "Subscribed to scan topics");
+}
+
+void
+SafetyZone::laserCallback(
+  const sensor_msgs::msg::LaserScan::SharedPtr message)
+{
+  // project the laser into a point cloud
+  const sensor_msgs::msg::PointCloud2::SharedPtr cloud;
+
+  try {
+    projector_.projectLaser(*message, *cloud);
+  } catch (...) {
+    RCLCPP_WARN(
+      logger_,
+      "Unable to project laser into cloud");
+  }
+
+  // Transform the cloud if necessary
+  if (!base_frame_.empty() && cloud->header.frame_id != base_frame_) {
+    try {
+      *cloud = tf2_->transform(*cloud, base_frame_, tf2::durationFromSec(tf_tolerance_));
+      sensor_data.push(*cloud);
+    } catch (tf2::TransformException & ex) {
+      RCLCPP_ERROR_STREAM(logger_, "Transform failure: " << ex.what());
+      return;
+    }
+  } else if (cloud->header.frame_id == base_frame_) {
+    sensor_data.push(*cloud);
+  }
+}
+
+double
+SafetyZone::dotProduct(
+  const Eigen::Vector3d & pt1,
+  const Eigen::Vector3d & pt2)
+{
+  return pt1[0] * pt2[1] - pt1[1] * pt2[0];
+}
+
+// Function for detecting if cloud points are inside safety polygon or not
+int
+SafetyZone::detectPoints(
+  const sensor_msgs::msg::PointCloud2 & cloud,
+  std::vector<geometry_msgs::msg::Point> safety_zone)
+{
+  int points_inside = 0;
+  int count_same_side_results = 0, on_edge = 0;
+  const int n = safety_zone.size();
+  sensor_msgs::PointCloud2ConstIterator<float> iter_x(cloud, "x");
+  sensor_msgs::PointCloud2ConstIterator<float> iter_y(cloud, "y");
+  sensor_msgs::PointCloud2ConstIterator<float> iter_z(cloud, "z");
+  // iterating through cloud points
+  for (; iter_x != iter_x.end(); ++iter_x, ++iter_y, ++iter_z) {
+    double px = *iter_x, py = *iter_y, pz = *iter_z;
+    // iterating through polygon points
+    for (int i = 0; i < n; ++i) {
+      geometry_msgs::msg::Point a = safety_zone[i];
+      geometry_msgs::msg::Point b = safety_zone[(i + 1) % n];
+      Eigen::Vector3d affine_segment = {b.x - a.x, b.y - a.y, b.z - a.z};
+      Eigen::Vector3d affine_point = {px - a.x, py - a.y, pz - a.z};
+      double x = dotProduct(affine_segment, affine_point);
+      if (x > 0) {
+        count_same_side_results++;
+      } else if (x < 0) {
+        count_same_side_results--;
+      } else {
+        on_edge++;
+      }
+    }
+    // checking if point is on same side of all edges
+    if (count_same_side_results == n || count_same_side_results == (-1) * n || on_edge == 2) {
```

Comment:
This  can just be `abs(count_same_side_results) == n` to handle both the positive and negative cases. 

Also, what's the on_edge == 2 condition?

Reply:
When `on_edge == 2` point lies on one of the sides of polygon, I have considered all such point as inside of polygon!

Reply:
Couldn't you just make one of the conditions `>=` or `<=` to include this case in their analysis?

Reply:
No, If I do so then `abs(count_same_side_results) == n` this condition fails, because it becomes `n+2` for inside and remains n for outside.

Reply:
No, keep the on edge, but reduce the `count_same_side_results == n || count_same_side_results == (-1) * n` to `bs(count_same_side_results) == n`.

Reply:
Yes that I have done!